### ISSUE TITLE
Firefox on android launch issues with android 11

### ISSUE THREAD

**Issue by sparsh00** [mozilla/geckodriver#1885]
## System

* Version: 0.29.1
* Platform: android 11
* Firefox: 88.1.1
* Selenium: appium 1.20.2

## Testcase

Trying to run session on devices like samsung galaxy s21 ultra (android 11) and get `adb error: fchown failed: Operation not permitted` with default androidStorage(sdcard) and `500 Internal Server Error {"value":{"error":"timeout","message":"Socket timeout reading Marionette handshake data: EOF reading marionette message","stacktrace":""}}` with androidStorage set to internal or app.

Note - Tried on pixel 4 android 11 and it runs fine (with default androidStorage and internal)

Had to add some extra logs in gd to get whats the exact issue
Logs androidStorage sdcard
[s21_sdcard_storage.log](https://github.com/mozilla/geckodriver/files/6472861/s21_sdcard_storage.log)

Logs androidStorage internal 
[s21_internal_storage.log](https://github.com/mozilla/geckodriver/files/6472863/s21_internal_storage.log)

Logs androidStorage app

Also in release note I can see that in future androidStorage will be deprecated (https://github.com/mozilla/geckodriver/releases/tag/v0.29.1). I think it can be a issue with devices with restriction on pushing data to directories like above

**Comment by whimboo** [comment]
Nothing in our code uses `fchown`. So what exactly are you trying to achieve here? Which changes did you make to geckodriver?

**Comment by sparsh00** [comment]
I just tried to log which file its not able to copy and got  `/sdcard/Android/data/org.mozilla.firefox/files/test_root/org.mozilla.firefox-geckodriver-profile/user.js` by adding `debug!("here6 {}", dest.display());` here https://searchfox.org/mozilla-central/source/testing/mozbase/rust/mozdevice/src/lib.rs#811
As per my debugging it fails while copying firefox profile https://searchfox.org/mozilla-central/source/testing/geckodriver/src/android.rs#325

**Comment by sparsh00** [comment]
Looks like a issue with Samsung and few other devices on android as can see a similar issue reported on [bundletool](https://github.com/google/bundletool) https://github.com/google/bundletool/issues/213. Handling in latest release - https://github.com/google/bundletool/compare/1.5.0...1.6.0#diff-a7c17bd9b0c6639db2f30e4e2e02c4e18185743ea84c8c16ec517f99d2a9fcadR219

**Comment by whimboo** [comment]
Thanks @sparsh00. Could you please check if `/data/local/tmp/` is writeable on your device?

**Comment by sparsh00** [comment]
@whimboo Yes it is.

**Comment by a-en** [comment]
hey,
seeing same issue on Android 9. Please find some details below:
geckodriver: 0.29.1
Android: 9
Firefox: 88.1.4
geckodriver start command: `geckodriver --android-storage sdcard`

Once geckodriver started, I'm trying to create a session by POST'ing to http://localhost:4444/session following payload:
```{
  "capabilities": {
    "alwaysMatch": {
      "moz:firefoxOptions": {
        "androidPackage": "org.mozilla.firefox",
        "androidDeviceSerial": "emulator-5554",
        "androidIntentArguments": [
          "-d", "http://example.org"
        ],
        "log": {"level": "trace"},
        "env": {
          "MOZ_LOG": "nsHttp:5",
          "MOZ_LOG_FILE": "/mnt/sdcard/log"
        }
      }
    }
  }
}
```

As result, Firefox gets opened in the emulator, but I got error

```
{
    "value": {
        "error": "timeout",
        "message": "Socket timeout reading Marionette handshake data: EOF reading marionette message",
        "stacktrace": ""
    }
}
```

Attaching [output](https://gist.github.com/a-en/d032fb8a11086c358e67d11f3ba3f6ab) from geckodriver as well


@whimboo could you take a look and let me know if I'm missing something?

**Comment by a-en** [comment]
Additional information, that might be helpful - I have tried 3 version of Firefox (release, beta, nightly) on Android 8.
The interesting thing - I've managed to create session for nightly Firefox by providing options mentioned [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1700290).

```
"androidIntentArguments": [
    "-a",
    "android.intent.action.VIEW",
    "-d",
    "about:blank",
]
```

Unfortunately, it doesn't work for the release and beta versions. But at least I have confirmation that everything seems to be OK with my setup, so the real issue is not my environment.

Version details:
Release - 88.1.4
Beta - 89.0.0-beta.10
Nightly - 90.0a1

**Comment by whimboo** [comment]
@a-en please remove all intent arguments from the capabilities. With geckodriver 0.29.1 default ones are used that will work for all kind of Firefox builds. And the ones you specified in your last comment are exactly those that are set as default now.

Can you please file a new issue if that still doesn't work? Your problem looks different through. Thanks a lot.

**Comment by whimboo** [comment]
> @whimboo Yes it is.

@sparsh00 so it means the following actually works on that Samsung device:

* Push the profile to `/data/local/tmp/%random%`.
* Move files on the device from `/data/local/tmp/%random%` to `/sdcard/Android/data/org.mozilla.firefox/files/test_root/org.mozilla.firefox-geckodriver-profile`
* Firefox uses this profile successfully
* Removing the above dirs is successful

**Comment by sparsh00** [comment]
@whimboo Yes it should works as you mentioned above as can push to `/data/local/tmp` and copy to `/sdcard/Android/data/org.mozilla.firefox/files/test_root`. 
Also i think this is what happens when we set `--android-storage = app`.

**Comment by whimboo** [comment]
Does it mean that `--android-storage=app` or another setting works for you?

I might have to patch geckodriver slightly so it runs a 2-step push. Lets see if I can find the time this week. 

But please let me know about my question. Thanks.

**Comment by sparsh00** [comment]
No `--android-storage=app` doesn't work, have attached logs for all in the issue description. Was just pointing out that have noticed the 2-step push for `app`.

**Comment by bdekoz** [comment]
somewhat off topic, but with --android-storage=app debug builds of fenix can be used with browsertime on android 11. Please don't delete this useful feature from geckodriver!

**Comment by whimboo** [comment]
@bdekoz why does the external storage option not work for debug builds? Can you please clarify? Thanks.

**Comment by stefanhuber** [comment]
I have the same problem and error message, with a Samsung S21 device and Android 12, Samsung S9 with Android 10 and same configuration works fine:

```
UnknownError: An unknown server-side error occurred while processing the command. Original error: adb error: fchown failed: Operation not permitted
```

with the following configuration for appium:

```
{
    'platformName': 'Windows',
    'appium:automationName': 'Gecko',
    'appium:deviceName': 'Samsung S21',
    'browserName': 'firefox',
    'acceptInsecureCerts': True,
    'moz:firefoxOptions': {
        'androidPackage': 'org.mozilla.firefox',
    }
}
```

**Are there any solutions to this problem yet?**

**Comment by whimboo** [comment]
@stefanhuber could you please attach trace logs and the ones from adb?

**Comment by stefanhuber** [comment]
@whimboo I have tried the exact same setup again and the issue dissapeared. So the issue is solved from my side...

**Comment by whimboo** [comment]
> @whimboo I have tried the exact same setup again and the issue dissapeared. So the issue is solved from my side...

So it means that it works now with the default setting of geckodriver to use the external sdcard?

**Comment by alyasporty** [comment]
@whimboo I am running into the same issue with Andriod>10 . I am not sure if it is related to the changes to the file permission https://developer.android.com/about/versions/11/privacy/storage

- Using latest driver 0.33.0
- Device tried Nexus and Pixel
- Tired `app:storage` auto, app, internal and sdcard and all produce the same error` adb error: fchown failed: operation not permitted`

**Comment by whimboo** [comment]
@alyasporty thanks for the report. I've filed https://bugzilla.mozilla.org/show_bug.cgi?id=1840443 to get this particular issue fixed. I can also replicate this failure when using a Pixel 6.

**Comment by 2343909** [comment]
@whimboo  I am getting the same error with Android 12 running on Samsung S22: " "adb error: fchown failed: Operation not permitted". The workaround using "--android-storage=app" does not work. 
Is there an estimate time as to when this issue will be resolved?

**Comment by whimboo** [comment]
> @whimboo I am getting the same error with Android 12 running on Samsung S22: " "adb error: fchown failed: Operation not permitted". The workaround using "--android-storage=app" does not work. Is there an estimate time as to when this issue will be resolved?

We don't have an ETA yet but we probably can take a look at it next month. Lets [follow-up on the above mentioned bug](https://github.com/mozilla/geckodriver/issues/1885#issuecomment-1607961971).

**Comment by whimboo** [comment]
I tried to reproduce locally with a Google Pixel phone but I cannot reproduce the problem. Could everyone who actually experience this issue tell us which kind of phone you are using and what the version of Android is? Thanks!

**Comment by 2343909** [comment]
@whimboo Hi, I am using a physical Samsung Galaxy S22 connected to the laptop with USB cable running with Android 12. All runs good on Chrome but Firefox responds with the noted error above.

**Comment by acreskeyMoz** [comment]
I'm seeing this error on a Samsung A54, straight from the factory (i.e. not rooted).

```
[2024-04-04 14:44:40] INFO: firefox failed to start, trying 0 more time(s): adb error: fchown failed: Operation not permitted
```

**Comment by ethanyou725** [comment]
same issue followed

**Comment by whimboo** [comment]
Hm, so everyone experiencing this issue so far seems to use a Samsung device for testing. I wonder if there is some locking mechanism that is preventing us from copying / moving data for the app on the device. Interesting is indeed that it works for chromedriver.

@sadym-chromium and @nechaev-chromium maybe you have a hint in our chromedriver copies profile data to the Android device, or don't you run such a step because all features are controllable via intent arguments?
